btrdbextras/opendss_ingest/simulation_utils.py

```diff
@@ -0,0 +1,569 @@
+            ncreated += 1
+    print("Found", nexisting, "streams. Created", ncreated, "streams.")
+    return existing
+
+
+def get_lineflow_stream_info(line_name, line_end, phase, ismag):
+    if ismag:
+        unit = "amps"
+    else:
+        unit = "degrees"
+    collection, name = get_lineflow_stream_colname(line_name, line_end, phase, ismag)
+
+    tags = {"name": name, "unit": unit}
+    annotations = {"phase": phase}
+
+    return collection, name, tags, annotations
+
+
+def get_lineflow_stream_colname(line_name, line_end, phase, ismag):
+    collection = line_name + "/" + line_end
+    if ismag:
+        lastltr = "M"
+    else:
+        lastltr = "A"
+
+    name = "I" + phase + lastltr
+    return collection, name
+
+
+def get_voltage_stream_column_name(bus_name, phase, ismag):
+    collection = bus_name
+    if ismag:
+        lastltr = "M"
+    else:
+        lastltr = "A"
+
+    name = "V" + phase + lastltr
+    return collection, name
+
+
+def get_voltage_stream_info(bus_name, phase, ismag, basekV):
+    if ismag:
+        unit = "volts"
+    else:
+        unit = "degrees"
+    collection, name = get_voltage_stream_column_name(bus_name, phase, ismag)
+    tags = {"name": name, "unit": unit}
+    annotations = {"phase": phase, "basekV": str(basekV)}
+
+    return collection, name, tags, annotations
+
+
+def add_all_data(times, data_dict, streams_dict, base_col):
+    """
+    Add data to each stream.
+
+    Parameters
+    ----------
+    times : list of ints
+        The timestamps for the data to be added (one set of times for all data)
+
+    data_dict : dict of arrays
+        The dictionary containing data to be added. Keys are the collection/name of
+        the stream to which data is to be added. Values are arrays of floats to add.
+
+    streams_dict : dict of stream objects
+        keys are the collection/name of each stream. values are the stream objects.
+
+    base_col : string
+        base collection prefix under which all streams can be found.
+    """
+    # Create progress bar
+    nstreams = len(data_dict)
+    pbar = tqdm(total=nstreams, desc="Pushing data to streams", leave=False)
+
+    for key, value in data_dict.items():
+        stream_info = f"{base_col}/{key}"
+        if stream_info in streams_dict:
+            add_to_stream(streams_dict[stream_info], times, value)
+        else:
+            print(f"WARNING: {stream_info} not found")
+        pbar.update(1)
+    pbar.close()
+
+
+def add_to_stream(stream, times, values):
+    """
+    Given times and values, put them in the required tuple format and
+    add them to the stream.
+    """
+    if len(times) != len(values):
+        print("WARNING: times & values not same size")
+
+    payload = list(zip(times, values))
+    stream.insert(payload, merge="replace")
+
+
+def ingest_streamset(streamset:StreamSet, data: DataFrame, period_ns: int):
+    while True:
+        for row in data.iterrows():
+            streamset.insert(row.to_dict())
```

Comment:
`iterrows` is terrifyingly slow sometimes, [10x slower than itertuples.](https://www.packetcoders.io/quickly-speed-up-your-panda-dataframes/) Can we use that here instead. The updated code would look something like this:
```suggestion
    while True:
        for row in data.itertuples(index=False):
            streamset.insert(row._asdict())
```